File: POH7/Megane/Main.erl
% Try POH
% author: Leonardone @ NEETSDKASU
-module(main).
-export([main/1,ngs/1,ngi/1,ngss/1,ngis/1]).
-import(io,[format/2]).
-import(lists,[seq/2,sublist/3,nthtail/2,takewhile/2,map/2]).

ti(S) -> {V,_} = string:to_integer(S), V.
gs() -> io:get_line("").
gi() -> ti(gs()).
gss() -> string:tokens(gs(), " ").
gis() -> lists:map(fun ti/1, gss()).
ngt(_,0) -> []; ngt(F,N) -> [F()|ngt(F,N-1)].
ngs(N) -> ngt(fun gs/0,N).
ngi(N) -> ngt(fun gi/0,N).
ngss(N) -> ngt(fun gss/0,N).
ngis(N) -> ngt(fun gis/0,N).

main(_) ->
    N = gi(),
    P = ngis(N),
    M = gi(),
    Q = ngis(M),
    T = length(takewhile(fun(E) -> E /= Q end, subpics(N,M,P))),
    NM = N - M + 1,
    Y = T div NM,
    X = T rem NM,
    format("~w ~w\n",[Y,X]),
    halt().

xparts(_,_,[]) -> [];
xparts(M,P,[X|XS]) -> [cutpic(X,M,P)|xparts(M,P,XS)].

subpics(_,M,P) when length(P) < M -> [];
subpics(N,M,P) -> xparts(M,P,seq(1,N-M+1)) ++ subpics(N,M,nthtail(1,P)).

cutpic(I,M,P) -> map(fun(E) -> sublist(E,I,M) end, sublist(P,1,M)).

File: POH7/Megane/Main.R
# Try POH
# author: Leonardone @ NEETSDKASU
# ==================================================================
rd <- local({
    lines <- readLines("stdin")
    pos <- 0
    gs <- function() { pos <<- pos + 1; lines[pos] }
    gi <- function() as.integer(gs())
    gss <- function() unlist(strsplit(gs()," "))
    gis <- function() as.integer(gss())
    ngt <- function(n,f) lapply(seq(n), function(x) f())
    ngs <- function(n) ngt(n,gs)
    ngi <- function(n) ngt(n,gi)
    ngss <- function(n) ngt(n,gss)
    ngis <- function(n) ngt(n,gis)
    scn <- function() {
        cur <- vector()
        s <- function() { if(is.na(cur[1])) cur <<- gss(); r <- cur[1]; cur <<- cur[2:length(cur)]; return(r) }
        i <- function() as.integer(s())
        ln <- function() { if(is.na(cur[1])) return(gs()); r <- paste(cur,collapse=" "); cur <<- vector(); return(r) }
        list(s=s, i=i, ln=ln)
    }
    list(gs=gs, gi=gi, gss=gss, gis=gis, ngs=ngs, ngi=ngi, ngss=ngss, ngis=ngis, scn=scn)
})
# funprogって項目にReduce,Filter,Find,Map,Positionがあってvectorの操作できる
# search() # ロードされてるパッケージを表示
# installed.packages() # パッケージを探す
# library(compiler); fc <- cmpfun(f) # 関数をコンパイルして高速にする
# ==================================================================

n <- rd$gi()
p <- rd$ngis(n)
m <- rd$gi()
q <- unlist(rd$ngis(m))

for(i in 1:(n-m+1)) {
    for(j in 1:(n-m+1)) {
        t <- unlist(Map(function(v) v[j:(j+m-1)], p[i:(i+m-1)]))
        if(all(t == q)) {
            cat(sprintf("%d %d\n", i - 1, j - 1))
        }
    }
}

